jwst/rscd/rscd_sub.py

```diff
@@ -1,15 +1,124 @@
 #
-#  Module for applying the RSCD correction to science data
+#  Module for the RSCD correction for MIRI science data
 #
 
 import numpy as np
 import logging
-from .. import datamodels
+from ..datamodels import dqflags
 
 log = logging.getLogger(__name__)
 
 
-def do_correction(input_model, rscd_model):
+def do_correction(input_model, rscd_model, type):
+    """
+    Short Summary
+    -------------
+    if type = baseline the correction sets initial groups in the integration
+    to skip
+    if type = enhanced the correction corrects the groups using a decay function
+
+    Parameters
+    ----------
+    input_model: data model object
+        science data to be corrected
+
+    rscd_model: rscd model object
+        rscd reference data
+
+    type: string
+        type of algorithm ['baseline' or 'enhanced']
+
+    Returns
+    -------
+    output_model: data model object
+        RSCD-corrected science data
+
+    """
+
+    output = input_model.copy()
+
+    # Retrieve the reference parameters for this exposure type
+    param = {}
+    param = get_rscd_parameters(input_model, rscd_model)
+
+    if param is None:
+        log.warning('RSCD correction will be skipped')
+        output.meta.cal_step.rscd = 'SKIPPED'
+        return output
+
+    if type == 'baseline':
+        group_skip = param['skip']
+        output = correction_skip_groups(input_model, group_skip)
+    else:
+        # enhanced algorithm is not enabled yet (updated code and validation needed)
+        log.warning('Enhanced algorithm not support yet: RSCD correction will be skipped')
+        output.meta.cal_step.rscd = 'SKIPPED'
+        return output
+        # decay function algorithm update needed
+        #output = correction_decay_function(input_model, param)
+
+    return output
+
+
+def correction_skip_groups(input_model, group_skip):
+    """
+    Short Summary
+    -------------
+    Set initial groups in integration to DO_NOT_USE to skip groups
+    affected by RSCD effect
+
+    Parameters
+    ----------
+    input_model: data model object
+        science data to be corrected
+
+    group_skip: int
+        number of groups to skip at the beginnig of the ramp
+
+    Returns
+    -------
+    output_model: data model object
+        RSCD-corrected science data
+    """
+
+    # Save some data params for easy use later
+    sci_nints = input_model.data.shape[0]       # number of integrations
+    sci_ngroups = input_model.data.shape[1]     # number of groups
+
+    log.debug("RSCD correction using: nints=%d, ngroups=%d" %
+              (sci_nints, sci_ngroups))
+
+    # Create output as a copy of the input science data model
+    output = input_model.copy()
```

Comment:
Wait, another instance of creating a copy of the input? Why is this done twice? One of them needs to be removed.

Reply:
Good point.  Made it `output = input_model` to avoid the copy.

Reply:
Ok?

Reply:
Nope.  Tests show that this doesn't work.  Keeping the copy in as otherwise the return has not been updated.  A bit confused here and happy to educated.

Reply:
It might be due to the fact that even though a copy is made up in the calling routine to create `output`, it's still `input_model` that's passed down to this routine. So another copy is needed down here. I'm guessing it's actually the copy in the calling routine that could be eliminated.

Reply:
Ah, now I get it.  I've removed the copy from the calling routine.